spec/support/helpers.rb

```diff
@@ -0,0 +1,151 @@
+require "tmpdir"
+require "open3"
+
+module Spec
+  module Helpers
+    def self.bundler_version
+      return @bundler_version if defined?(@bundler_version)
+      versions = Bundler.with_clean_env do
+        `gem list bundler`.lines.grep(/^bundler /).first.scan(/\d+\.\d+\.\d+/)
+      end
+      to_find = ENV["TEST_BUNDLER_VERSION"] || ENV["BUNDLER_VERSION"]
+      @bundler_version = versions.detect { |v| v.include?(to_find) }
+      raise "Unable to find bundler version: #{to_find.inspect}" if @bundler_version.nil?
+      @bundler_version
+    end
+
+    def self.bundler_short_version
+      bundler_version.rpartition(".").first
+    end
+
+    # Ruby gem binstubs allow you to pass a secret version in order to load a
+    # bin file for a particular gem version. This is useful when you have
+    # multiple versions of a gem installed and only want to invoke a specific
+    # one.
+    #
+    # So, if I have bundler 1.17.3 and 2.0.1 installed, I can run 1.17.3 with:
+    #
+    #     bundle _1.17.3_ update
+    #
+    def self.bundler_cli_version
+      "_#{bundler_version}_"
+    end
+
+    def self.global_bundler_d_dir
+      @global_bundler_d_dir ||= Pathname.new("~/.bundler.d").expand_path
+    end
+
+    def self.global_bundler_d_backup_dir
+      @global_bundler_d_backup_dir ||= Pathname.new("~/.bundler.d.rspec_backup").expand_path
+    end
+
+    def self.backup_global_bundler_d
+      FileUtils.rm_rf(global_bundler_d_backup_dir)
+      FileUtils.mv(global_bundler_d_dir, global_bundler_d_backup_dir) if global_bundler_d_dir.exist?
+    end
+
+    def self.restore_global_bundler_d
+      FileUtils.rm_rf(global_bundler_d_dir)
+      FileUtils.mv(global_bundler_d_backup_dir, global_bundler_d_dir) if global_bundler_d_backup_dir.exist?
+    end
+
+    attr_reader :out, :err, :process_status
+
+    def bundler_version
+      Helpers.bundler_version
+    end
+
+    def bundler_short_version
+      Helpers.bundler_short_version
+    end
+
+    def app_dir
+      @app_dir ||= Pathname.new(Dir.mktmpdir)
+    end
+
+    def rm_app_dir
+      return unless @app_dir
+      FileUtils.rm_rf(@app_dir)
+      @app_dir = nil
+    end
+
+    def rm_global_bundler_d_dir
+      FileUtils.rm_rf(Helpers.global_bundler_d_dir)
+    end
+
+    def with_path_based_gem(source_repo)
+      Dir.mktmpdir do |path|
+        path = Pathname.new(path)
+        Dir.chdir(path) { `git clone --depth 1 #{source_repo} the_gem 2>/dev/null` }
```

Comment:
Alternative:

```ruby
Open3.capture2e("git clone -C #{path} --depth 1 #{source_repo} the_gem")
```

(the `-C #{path}` is just a replacement for `Dir.chdir(path) {}`.  You could do either.)

Also, this method allows you to capture the output and if there was an error, you can raise it and print it as part of the error message.

Reply:
I like the catch an error and raise it...I'll do that